common/app-starters-task-launch-request-common/src/test/java/org/springframework/cloud/stream/app/tasklaunchrequest/TaskLaunchRequestIntegrationTests.java

```diff
@@ -0,0 +1,204 @@
+
+		try (ConfigurableApplicationContext context = new SpringApplicationBuilder(
+			TestChannelBinderConfiguration.getCompleteConfiguration(TestApp.class)).web(WebApplicationType.NONE)
+			.run("--spring.jmx.enabled=false",
+				"---task.launch.request.task-name=foo", "--enhanceTLRContext=true")) {
+
+			DataFlowTaskLaunchRequest dataFlowTaskLaunchRequest = verifyAndreceiveDataFlowTaskLaunchRequest(context);
+
+			assertThat(dataFlowTaskLaunchRequest.getTaskName()).isEqualTo("foo");
+			assertThat(dataFlowTaskLaunchRequest.getCommandlineArguments()).hasSize(1);
+			assertThat(dataFlowTaskLaunchRequest.getCommandlineArguments()).containsExactly("runtimeArg");
+		}
+	}
+
+	private DataFlowTaskLaunchRequest verifyAndreceiveDataFlowTaskLaunchRequest(ApplicationContext context)
+		throws IOException {
+		return this.verifyAndreceiveDataFlowTaskLaunchRequest(context, null);
+	}
+
+	private DataFlowTaskLaunchRequest verifyAndreceiveDataFlowTaskLaunchRequest(ApplicationContext context,
+		TaskLaunchRequestContext taskLaunchRequestContext) throws IOException {
+
+		MessageChannel input = context.getBean("input", MessageChannel.class);
+
+		PollableChannel output = context.getBean("output", PollableChannel.class);
+
+		MessageBuilder<byte[]> builder = MessageBuilder.withPayload(new byte[] {});
+
+		if (taskLaunchRequestContext != null) {
+			builder.setHeader(TaskLaunchRequestContext.HEADER_NAME, taskLaunchRequestContext);
+		}
+
+		input.send(builder.build());
+
+		Message<?> message = output.receive(1000);
+
+		assertThat(message).isNotNull();
+
+		return (DataFlowTaskLaunchRequest) message.getPayload();
+	}
+
+	@Configuration
+	@EnableAutoConfiguration
+	static class TestApp {
+		@Bean
+		public TaskLaunchRequestContextProvider taskLaunchRequestContextProvider(@Value("${enhanceTLRContext:false}")
+			boolean enhance) {
+			return new TaskLaunchRequestContextProvider(enhance);
+		}
+
+		@Bean
+		public MessageChannel input() {
+			return new DirectChannel();
+		}
+
+		@Bean
+		MessageChannel output() {
+			return new QueueChannel();
+		}
+
+		@Autowired
+		TaskLaunchRequestTransformer taskLaunchRequestTransformer;
+
+		@Bean
+		public IntegrationFlow flow(TaskLaunchRequestContextProvider provider) {
+
+			return IntegrationFlows.from(input()).transform(provider).transform(taskLaunchRequestTransformer)
+				.channel(output()).get();
+		}
+
+		static class TaskLaunchRequestContextProvider implements MessageProcessor<Message> {
+
+			private final boolean enhance;
+
+			TaskLaunchRequestContextProvider(boolean enhance) {
+				this.enhance = enhance;
+			}
+
+			@Override
+			public Message<?> processMessage(Message<?> message) {
+				if (enhance) {
+
+					TaskLaunchRequestContext taskLaunchRequestContext = new TaskLaunchRequestContext();
+					taskLaunchRequestContext.addCommandLineArg("runtimeArg");
+
+					message = MessageBuilder.fromMessage(message).setHeader(TaskLaunchRequestContext.HEADER_NAME,
+						taskLaunchRequestContext).build();
```

Comment:
This is not good in case of `transform()`. Since you return the whole `Message`, it is not going to copy request headers.
I would suggest to go an `.enrichHeaders()` way:
```
.enrichHeaders(h -> h.headerFunction(TaskLaunchRequestContext.HEADER_NAME, 
                                                                 m -> {
                                        TaskLaunchRequestContext taskLaunchRequestContext = new TaskLaunchRequestContext();
					taskLaunchRequestContext.addCommandLineArg("runtimeArg");
                                        return taskLaunchRequestContext;
                                                                          }))
```